### PR TITLE
Handle 'reduce' of a forall of an iterable

### LINKED ISSUE #16027: compiler assertion error due to reduction of forall expression of promoted expression
### Summary of Problem

When reducing a forall loop whose body is a promoted expression (e.g., `[i in 1..10] i * Arr`), we hit a compiler assertion error (`internal error: RES-FUN-ION-10082`) due to the `chpl__sumType()` routine in ChapelReduce.chpl apparently trying to create a default-initialized variable of the iterator expression's type (issue #16026).  I believe the reduction itself ought to be legal and that the quick fix would be to have `chpl__sumType()` handle the case of an iterator expression more explicitly / precisely rather than solving #16026.

### Steps to Reproduce

**Associated Future Test(s):**
[`test/reductions/reduceLoopOfPromoted.chpl`](
           https://github.com/chapel-lang/chapel/blob/master/test/reductions/reduceLoopOfPromoted.chpl
          ) #16025

### Configuration Information

- Output of `chpl --version`: `chpl version 1.23.0 pre-release (1032009515)`

### PR DESCRIPTION
**Pull request by vasslitvinov**
Resolves #16027, handling reduce expressions of this kind:

```chpl
+ reduce ([outerIdx in outerShape] (inner iterable expression))
```

The solution here is to have the accumulation state be an array of the same shape as the inner iterable.

The tricky part of the implementation is to determine the shape of the inner iterable. For that, we fire up another loop over `outerShape`, breaking out of it once we get the first yielded value.

While there, change a compiler internal error to a user error for #16462 and add tests for #16546.

### CODE REVIEW

**Comment by vasslitvinov** [comment]
@bradcray @mppf it took a bit more complexity to get #16027's example to work than just fixing a compiler assert.

My concern is that advancing the outer iterable might lead to unexpected effects including performance and extra execution. Although in the common case where outerShape is an array or a domain or another domain-shaped iterable, this should not be a concern.

Yet, I propose instead to have the user do this transformation. That is, instead of:
```chpl
+ reduce ([outerIdx in outerShape] (inner iterable expression))
```
have the user write:
```chpl
var result = [the domain of the inner iterable] elementType;
forall outerIdx in outerShape with (+ reduce result) do
  result reduce= inner iterable;
```

For that, I propose for the compiler to issue a user error when it detects such a `reduce` expression. The error would indicate that this is not implemented and request a rewrite as a forall loop.

Thoughts?

**Comment by bradcray** [comment]
I think that generating a user-facing error for cases like this as a stopgap rather than an internal error would be a step forward.  But ultimately, I think that it would be surprising if the language did not support reductions like `var z = + reduce ([i in 1..3] (i * A));`.

**Comment by mppf** [comment]
@vasslitvinov - it seems to me that we need #5578 or something along those lines and that would naturally fit into implementing the iterator as a record (e.g. there could be a method to get the yielded type). I've added a note in #11505.

File: modules/internal/ChapelIteratorSupport.chpl [hunk 1]
```diff
@@ -104,6 +104,51 @@ module ChapelIteratorSupport {
...
+    // is an array and therefore has a runtime component.
+
+    for i in x {
+      compilerAssert(i.type <= _iteratorRecord); // prevent unintended use
+
+      // The case where the shape is a range can be implemented
+      // analogously to the two versions of chpl__initCopy(_iteratorRecord)
+      // that invoke chpl__initCopy_shapeHelp().
+      if !chpl_iteratorHasDomainShape(i) then
+        compilerError("unsupported elements of the expression being reduced -- they are iterable expressions without a domain shape");
```
Comment (mppf) [inline modules/internal/ChapelIteratorSupport.chpl:133]: avoid the too-long line by breaking apart the error with `compilerError("part1 " + "part2 ")` with the two literals on different lines

File: modules/internal/ChapelIteratorSupport.chpl [hunk 2]
```diff
@@ -104,6 +104,51 @@ module ChapelIteratorSupport {
...
+
+      // Part 2 - build the array type -- as in chpl__initCopy(_iteratorRecord)
+      var shape = new _domain(i._shape_);
+      shape._unowned = true;
+
+      type arrElt = iteratorIndexType(i);
+      if isArray(arrElt) || isDomain(arrElt) then
+        // This scenario needs testing esp. memory/leaks before enabling it.
+        compilerError("unsupported elements of the expression being reduced -- they are iterable expressions consisting of arrays or domains");
```
Comment (mppf) [inline modules/internal/ChapelIteratorSupport.chpl:142]: same here

**Review by mppf — APPROVED** [review]
Could you add a test with double nesting, like 
``` chapel
var z = + reduce ([i in 1..3] [j in 1..3] (10 * A));
```
?

**Pull request merged**